relay-metrics/src/protocol.rs

```diff
@@ -390,47 +467,57 @@ pub struct Metric {
 impl Metric {
     /// Creates a new metric using the MRI naming format.
     ///
+    /// See [`Metric::name`].
+    ///
     /// MRI is the metric resource identifier in the format `<type>:<ns>/<name>@<unit>`. This name
     /// ensures that just the name determines correct bucketing of metrics with name collisions.
     pub fn new_mri(
-        namespace: impl fmt::Display,
+        namespace: MetricNamespace,
         name: impl fmt::Display,
         unit: MetricUnit,
         value: MetricValue,
         timestamp: UnixTimestamp,
         tags: BTreeMap<String, String>,
     ) -> Self {
         Self {
-            name: format!("{}:{}/{}@{}", value.ty(), namespace, name, unit),
-            unit,
+            name: MetricResourceIdentifier {
+                ty: value.ty(),
+                name: name.to_string().into(),
+                namespace,
+                unit,
+            }
+            .to_string(),
             value,
             timestamp,
             tags,
         }
     }
 
+    /// Parse statsd-compatible payload of format
+    /// ```text
+    /// [<ns>/]<name>[@<unit>]:<value>|<type>[|#<tags>]`
+    /// ```
     fn parse_str(string: &str, timestamp: UnixTimestamp) -> Option<Self> {
-        let mut components = string.split('|');
-
-        let name_value_str = components.next()?;
-        let ty = components.next().and_then(|s| s.parse().ok())?;
-        let (name, unit, value) = parse_name_unit_value(name_value_str, ty)?;
+        let (raw_namespace, rest) = string.split_once('/').unwrap_or(("custom", string));
```

Comment:
I think you should keep the previous structure here, of primarily splitting by `|` and only then parse the name out of the `name_value_str`. Otherwise, a slash in a tag (which could be permitted) would break this.

Ideally, we have a reusable internal function to parse the portion of the MRI without the type prefix that's both used here and in the `FromStr for Mri` implementation.